libp2p-secio/src/handshake.rs

```diff
@@ -0,0 +1,752 @@
+
+		// Check that the received nonce is correct.
+		.and_then(|(codec, context)| {
+			codec.into_future()
+				.map_err(|(e, _)| e)
+				.and_then(move |(nonce, rest)| {
+					match nonce {
+						Some(ref n) if n == &context.local_nonce => {
+							Ok((rest, context.remote_public_key))
+						},
+						None => {
+							Err(IoError::new(IoErrorKind::BrokenPipe, "unexpected eof").into())
+						},
+						_ => Err(SecioError::NonceVerificationFailed)
+					}
+				})
+		});
+
+	Box::new(future)
+}
+
+// Custom algorithm translated from reference implementations. Needs to be the same algorithm
+// amongst all implementations.
+fn stretch_key(key: &SigningKey, result: &mut [u8]) {
+	const SEED: &'static [u8] = b"key expansion";
+
+	let mut init_ctxt = SigningContext::with_key(key);
+	init_ctxt.update(SEED);
+	let mut a = init_ctxt.sign();
+
+	let mut j = 0;
+	while j < result.len() {
+		let mut context = SigningContext::with_key(key);
+		context.update(a.as_ref());
+		context.update(SEED);
+		let b = context.sign();
+
+		let todo = cmp::min(b.as_ref().len(), result.len() - j);
+
+		result[j..j + todo].copy_from_slice(&b.as_ref()[..todo]);
+
+		j += todo;
+
+		let mut context = SigningContext::with_key(key);
+		context.update(a.as_ref());
+		a = context.sign();
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::handshake;
+	use super::stretch_key;
+	use futures::Future;
+	use futures::Stream;
+	use ring::digest::SHA256;
+	use ring::hmac::SigningKey;
+	use ring::signature::RSAKeyPair;
+	use std::sync::Arc;
+	use tokio_core::net::TcpListener;
+	use tokio_core::net::TcpStream;
+	use tokio_core::reactor::Core;
+	use untrusted::Input;
+
+	#[test]
+	fn handshake_with_self_succeeds() {
+		let mut core = Core::new().unwrap();
+
+		let private_key1 = {
+			let pkcs8 = include_bytes!("../tests/test-private-key.pk8");
+			Arc::new(RSAKeyPair::from_pkcs8(Input::from(&pkcs8[..])).unwrap())
+		};
+		let public_key1 = include_bytes!("../tests/test-public-key.der").to_vec();
+
+		let private_key2 = {
+			let pkcs8 = include_bytes!("../tests/test-private-key-2.pk8");
+			Arc::new(RSAKeyPair::from_pkcs8(Input::from(&pkcs8[..])).unwrap())
+		};
+		let public_key2 = include_bytes!("../tests/test-public-key-2.der").to_vec();
+
+		let listener = TcpListener::bind(&"127.0.0.1:0".parse().unwrap(), &core.handle()).unwrap();
+		let listener_addr = listener.local_addr().unwrap();
+
+		let server = listener.incoming()
+							 .into_future()
+							 .map_err(|(e, _)| e.into())
+							 .and_then(move |(connec, _)| {
+			handshake(connec.unwrap().0, public_key1, private_key1)
+		});
+
+		let client = TcpStream::connect(&listener_addr, &core.handle())
+			.map_err(|e| e.into())
+			.and_then(move |stream| handshake(stream, public_key2, private_key2));
+
+		core.run(server.join(client)).unwrap();
+	}
+
+	#[test]
+	fn stretch() {
+		let mut output = [0u8; 32];
+
+		let key1 = SigningKey::new(&SHA256, &[]);
+		stretch_key(&key1, &mut output);
+		assert_eq!(
+			&output,
+			&[
```

Comment:
Sorry for the very bad formatting, but I think [the option to fix that in rustfmt doesn't work](https://github.com/rust-lang-nursery/rustfmt/blob/master/Configurations.md#array_horizontal_layout_threshold).

`rustfmt` also likes to convert tabs into spaces for some reason, but only at some specific places.